src/librustc_mir/interpret/snapshot.rs

```diff
@@ -46,361 +38,338 @@ pub(crate) struct InfiniteLoopDetector<'mir, 'tcx> {
+}
+
+impl InterpRef<'a, 'mir, 'tcx> {
+    fn new(stack: &'a [Frame<'mir, 'tcx>], memory: &'a CtfeMemory<'mir, 'tcx>) -> Self {
+        InterpRef {
+            stack,
+            memory,
         }
     }
-}
 
-impl_snapshot_for!(enum Immediate {
-    Scalar(s),
-    ScalarPair(s, t),
-});
-
-impl_snapshot_for!(enum Operand {
-    Immediate(v),
-    Indirect(m),
-});
-
-impl_snapshot_for!(enum LocalValue {
-    Dead,
-    Uninitialized,
-    Live(v),
-});
-
-impl<'a, Ctx> Snapshot<'a, Ctx> for Relocations
-    where Ctx: SnapshotContext<'a>,
-{
-    type Item = Relocations<(), AllocIdSnapshot<'a>>;
-
-    fn snapshot(&self, ctx: &'a Ctx) -> Self::Item {
-        Relocations::from_presorted(self.iter()
-            .map(|(size, ((), id))| (*size, ((), id.snapshot(ctx))))
-            .collect())
+    /// Returns an iterator over all memory allocations reachable from anywhere on the stack.
+    ///
+    /// The order of iteration is predictable: Calling this method twice on a single
+    /// `InterpRef` will yield the same `Allocation`s in the same order.
+    fn live_allocs(
+        &self,
+    ) -> memory::DepthFirstSearch<'a, 'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>> {
+        let stack_roots = self.stack
+            .iter()
+            .flat_map(|frame| frame.locals.iter())
+            .filter_map(|local| local.operand())
+            .flat_map(|op| op.alloc_ids().into_iter());
+
+        memory::DepthFirstSearch::with_roots(self.memory, stack_roots)
     }
 }
 
-#[derive(Eq, PartialEq)]
-struct AllocationSnapshot<'a> {
-    bytes: &'a [u8],
-    relocations: Relocations<(), AllocIdSnapshot<'a>>,
-    undef_mask: &'a UndefMask,
-    align: &'a Align,
-    size: &'a Size,
-    mutability: &'a Mutability,
-}
+impl PartialEq for InterpRef<'_, 'mir, 'tcx> {
+    fn eq(&self, other: &Self) -> bool {
+        let is_stack_eq = self.stack.iter().map(AllocIdInvariant)
+            .eq(other.stack.iter().map(AllocIdInvariant));
 
-impl<'a, Ctx> Snapshot<'a, Ctx> for &'a Allocation
-    where Ctx: SnapshotContext<'a>,
-{
-    type Item = AllocationSnapshot<'a>;
+        if !is_stack_eq {
+            return false;
+        }
 
-    fn snapshot(&self, ctx: &'a Ctx) -> Self::Item {
-        let Allocation {
-            size,
-            align,
-            mutability,
-            extra: (),
-            ..
-        } = self;
+        let live_allocs = self.live_allocs();
+        let other_live_allocs = other.live_allocs();
 
-        let all_bytes = 0..self.len();
-        // This 'inspect' is okay since following access respects undef and relocations. This does
-        // influence interpreter exeuction, but only to detect the error of cycles in evalution
-        // dependencies.
-        let bytes = self.inspect_with_undef_and_ptr_outside_interpreter(all_bytes);
+        live_allocs.map(|(_, alloc)| alloc.ok().map(AllocIdInvariant))
+            .eq(other_live_allocs.map(|(_, alloc)| alloc.ok().map(AllocIdInvariant)))
```

Comment:
@RalfJung. I discard the contents of the `InterpError` with `ok`. Do you think we can differentiate between null pointers, pointers to ZSTs and dangling pointers by looking at `InterpError` like you mentioned above? Do ZST pointers have zero-sized backing allocations?